File: generadorLineal_p4.py
"""
Generador de números pseudoaleatorios
Fuente: Dromey, 'How to solve it by computer'
Algorithm 3.6, 'Generation of pseudo-random numbers'
Adaptado para Python por Ignacio Trejos Zelaya
Modificado por Alonso Garita Granados
"""

import generadorPrimos as gp

def aleatorio (semilla, n):
    ## x tomará como base la semilla
    ## n es la cantidad de números pseudoaleatorios que deseamos generar

    if(semilla < 11): #Se asegura que la semilla sea al menos 11
        semilla = 11

    ## Inicializaciones
    ## La literatura  explica la importancia de elegir bien los valores m, a y b
    m = 4096 ## período
    a = 109  ## multiplicador
    b = 853  ## incremento

    ## x será la semilla para iterar (esto puede ser modificado)

    x = 0
    for p in gp.generarPrimos(semilla) : #Hasta que x >= semilla no sale del ciclo
        if(p >= semilla):
            x = p
            break
    #Cuando sale, con certeza x >= semilla

    for i in range (n):
        ## calcular el siguiente número aleatorio
        x = (a * x + b) % m

        ## devolver el siguiente número aleatorio
        yield x

    ## terminar (el for que consuma a este iterador nunca llegará aquí)
    return "fin"

## Semilla, cantidad, rango, frecuencias
semilla = int(input ("Ingrese una semilla, por favor: "))
cuántos = int(input ("¿Cuántos números pseudo-aleatorios desea generar?: "))
rango = int(input ("¿Cuál es el rango de los valores por generar (0 hasta r)?: "))
rango += 1 ## ajuste al rango

lista = [ r for r in aleatorio (semilla, cuántos) ]

for r in lista:
        print (r)

en_rango = [ r % rango for r in aleatorio (semilla, cuántos) ]

for r in en_rango:
        print (r)

File: Lua/generadorLineal_p4.lua

-- Generador de números pseudoaleatorios
-- Fuente: Dromey, 'How to solve it by computer'
-- Algorithm 3.6, 'Generation of pseudo-random numbers'
-- Adaptado para Python por Ignacio Trejos Zelaya
-- Modificado por Alonso Garita Granados
-- Implementado en Lua por Raquel Arguedas
-- Fecha de inicio 15/09/2022
-- Fecha de finalización 20/09/2022

co = coroutine.create(
	function (semilla, n)
		-- si la semilla es menor a 11 la convierte en 11
		if(semilla < 11)then 
        	semilla = 11
		end

		-- La literatura  explica la importancia de elegir bien los valores m, a y b
		m = 4096 -- período
    	a = 109  -- multiplicador
    	b = 853  -- incremento
		
		-- busca el próximo primo
		x = semilla
		flag = 1
		while (flag == 1) do
			flag = 0
			for var=2,9 do
   				if x%var == 0 then
					flag = 1
					x = x + 1
				end
				if flag == 1 then break end
			end
		end 

		for i=0,n-1 do
			-- devolver el siguiente número aleatorio
			coroutine.yield(x)

        	-- calcular el siguiente número aleatorio
        	x = (a * x + b) % m
		end
	end
)

-- para realizar pruebas
n=5
semillaX=555

for i=0,n-1 do
	isAlive, value = coroutine.resume(co, semillaX, n)
	print(value)
end

